### PR TITLE
Add failure mode to `write_list`

### PR DESCRIPTION
**Pull request by marcusklaas**
Fixes some cases where we would exceed the column limit.

### CODE REVIEW

File: tests/target/paths.rs [hunk 1]
```diff
@@ -1,16 +1,16 @@
...
+        Constellation::<layout::layout_task::LayoutTask,
+                        script::script_task::ScriptTask>::start(compositor_proxy,
+                                                                resource_task,
+                                                                image_cache_task,
+                                                                font_cache_task,
+                                                                time_profiler_chan,
+                                                                mem_profiler_chan,
+                                                                devtools_chan,
+                                                                storage_task,
+                                                                supports_clipboard);
```
Comment (nrc) [inline tests/target/paths.rs]: nice :-)

File: src/utils.rs [hunk 2]
```diff
@@ -189,6 +189,40 @@ macro_rules! try_opt {
     })
 }
 
+pub fn wrap_str<S>(s: S, max_width: usize, width: usize, offset: usize) -> Option<String>
+    where S: Into<String>
```
Comment (nrc) [inline src/utils.rs]: I think it might be better to take an `&str` rather than use `Into<String>` - it will save an allocation in some cases.

File: src/utils.rs [hunk 3]
```diff
@@ -189,6 +189,40 @@ macro_rules! try_opt {
     })
 }
 
+pub fn wrap_str<S>(s: S, max_width: usize, width: usize, offset: usize) -> Option<String>
```
Comment (nrc) [inline src/utils.rs]: Could you document this fn please

File: src/expr.rs [hunk 4]
```diff
@@ -854,19 +829,58 @@ fn rewrite_call(context: &RewriteContext,
...
+            Err(Ordering::Less) => {
+                lo = middle;
+                tries += 1;
+            }
+            Err(Ordering::Greater) => {
+                hi = middle;
+                tries += 1;
+            }
+            _ => unreachable!(),
+        }
```
Comment (nrc) [inline src/expr.rs]: I would try to factor out the abstract searching from the concrete reformatting here (and add unit tests for the search, binary search is notorious for off-by-one errors etc.)

File: src/expr.rs [hunk 5]
```diff
@@ -854,19 +829,58 @@ fn rewrite_call(context: &RewriteContext,
...
-    let callee_str = try_opt!(callee.rewrite(context, max_callee_width, offset));
-    debug!("rewrite_call, callee_str: `{}`", callee_str);
-
-    if args.is_empty() {
-        return Some(format!("{}()", callee_str));
+    let mut hi = try_opt!(width.checked_sub(2)) * 2;
+    let mut lo = 0;
+    let mut tries = 0;
+
+    // Binary search for the best split between callee and arguments.
```
Comment (nrc) [inline src/expr.rs]: Could you add to the comment: what exactly are you trying to optimise? In what cases will this succeed/fail on the first attempt? Do you ever expect it to take >10 tries or is that just for safety? (btw 10 seems pretty pessimistic, given that there is only a 100 chars max, I'd expect to get a decent result in something like 2 or 3 tries or never (although, I don't quite understand completely what is going on here). Also, is it possible to go with a 'reasonable' answer after n tries, rather than giving up completely if we don't get a perfect answer?

Comment (marcusklaas) [inline src/expr.rs]: It actually takes the first reasonable answer it can get. 10 was taken as a safety measure, indeed. 2log(10) should be better. I'll be sure to do the refactor and add documentation.

File: src/expr.rs [hunk 6]
```diff
@@ -1191,12 +1211,13 @@ pub fn rewrite_assign_rhs<S: Into<String>>(context: &RewriteContext,
...
+            // where is the 1 coming from anyway?
             let max_width = try_opt!(context.config.max_width.checked_sub(new_offset + 1));
-            let rhs = try_opt!(ex.rewrite(&context.overflow_context(context.config.tab_spaces),
-                                          max_width,
-                                          new_offset));
+            let overflow_context = context.overflow_context(context.config.tab_spaces);
+            let rhs = ex.rewrite(&overflow_context, max_width, new_offset);
 
-            result.push_str(&rhs);
+            result.push_str(&&try_opt!(rhs));
```
Comment (nrc) [inline src/expr.rs]: does this really need another `&`?

Comment (marcusklaas) [inline src/expr.rs]: Did you forget? :-P

https://github.com/nrc/rustfmt/issues/172

Comment (nrc) [inline src/expr.rs]: Oh god, yeah, that stupid thing

**Comment by nrc** [comment]
looks good! Other than the inline comments, there are a couple of FIXMEs on magic `100`s. I think you should fix them now, since I don't think they will get any easier in the future and are the kind of thing that will sit around without a fix for a long time, but cause some subtle bug when someone actually tries running with a non-100 max width. r+ with those fixed.

File: src/expr.rs [hunk 7]
```diff
@@ -854,19 +829,58 @@ fn rewrite_call(context: &RewriteContext,
...
+
+    // Binary search for the best split between callee and arguments.
+    loop {
+        let middle = (lo + hi) / 2;
+
+        match rewrite_call_inner(context, callee, middle, args, span, width, offset) {
+            _ if tries > 10 => return None,
+            Ok(result) => return Some(result),
+            Err(Ordering::Less) => {
+                lo = middle;
```
Comment (sinhpham) [inline src/expr.rs]: This may be way off, but should it be ```lo = middle + 1``` to avoid infinite loop in case hi = lo + 1? Same for ```hi = middle - 1```

Comment (marcusklaas) [inline src/expr.rs]: You're absolutely right! I guess this proves @nrc's point about binary search being prone to off-by-one errors ;-)
